File: extra/scripts/dff.py
#!/usr/bin/env python3

from typing import List, Optional
from enum import Enum
from dataclasses import dataclass
import subprocess
import re
from rich.table import Table
from rich.console import Console

RELEVANT_MOUNT_POINT_RE = re.compile("/($|Volumes|media|usr|mnt)")
SPACES_RE = re.compile("\\s+")


class State(Enum):
    OK = "OK"
    GOOD = "GOOD"
    BAD = "BAD"


@dataclass
class Disk:
    device: str
    used: str
    used_percentage: int
    size: int
    available: int
    mount: str

    def size_gb(self) -> str:
        return size_to_gb(self.size)

    def available_gb(self) -> str:
        return size_to_gb(self.available)

    def state(self) -> State:
        if self.used_percentage > 90:
            return State.BAD
        elif self.used_percentage > 70:
            return State.OK
        return State.GOOD

    def style(self) -> str:
        state = self.state()
        if state == State.BAD:
            return "red"
        elif state == State.OK:
            return "orange"
        return "green"


def main():
    render_disks(load_disks())


def render_disks(disks: List[Disk]):
    table = Table(title="Free Space", border_style="#666666")
    table.add_column("Mount")
    table.add_column("Usage", justify="right")
    table.add_column("Free (GB)", justify="right")
    table.add_column("Size (GB)", justify="right")
    table.add_column("Device")

    for disk in disks:
        table.add_row(
            disk.mount,
            f"{disk.used_percentage}%",
            disk.available_gb(),
            disk.size_gb(),
            disk.device,
            style=disk.style(),
        )

    console = Console()
    console.print(table)


def load_disks() -> List[Disk]:
    lines = subprocess.check_output(["df"]).decode("utf-8").splitlines()[1:]
    disks = map(parse_df_line, lines)
    disks = [
        disk
        for disk in disks
        if disk is not None and RELEVANT_MOUNT_POINT_RE.match(disk.mount)
    ]
    return disks


def parse_df_line(line: str) -> Optional[Disk]:
    *device, size, used, available, used_percentage, mount = SPACES_RE.split(
        line.strip()
    )
    return Disk(
        device=" ".join(device),
        used=used,
        used_percentage=int(used_percentage[:-1]),
        size=int(size),
        available=int(available),
        mount=mount,
    )


def size_to_gb(value: int) -> str:
    return "%.1f G" % (value / 1024 / 1024)


if __name__ == "__main__":
    main()

File: extra/scripts/deno/dff.ts
#!/usr/bin/env -S deno run --allow-env --allow-read --allow-run

import { Table } from "npm:console-table-printer"

const SPACES_REGEX = /\s+/
const EXCLUDE_MOUNT_POINTS = /^\/(System\/Volumes|snap)\//
const MIN_FREE_GB = 4

interface Disk {
  device: string
  used: number
  usedPercentage: number
  size: number
  availableKB: number
  mount: string
}

function rowColor({ usedPercentage }: Disk): string {
  if (usedPercentage > 90) {
    return "red"
  }
  if (usedPercentage > 70) {
    return "yellow"
  }
  return "green"
}

function diskToRow(disk: Disk): Record<string, string> {
  const { device, usedPercentage, size, availableKB, mount } = disk
  return {
    mount,
    usage: `${usedPercentage}%`,
    free: prettifyKB(availableKB),
    size: prettifyKB(size),
    device,
  }
}

function main() {
  const disks = loadDisks()

  if (Deno.args.includes("-w") || Deno.args.includes("--warn")) {
    warn(disks)
    return
  }

  const table = new Table({
    columns: [
      {
        name: "mount",
        title: "Mount",
        alignment: "left",
      },
      {
        name: "usage",
        title: "Usage",
        alignment: "right",
      },
      {
        name: "free",
        title: "Free",
        alignment: "right",
      },
      {
        name: "size",
        title: "Size",
        alignment: "right",
      },
      {
        name: "device",
        title: "Device",
        alignment: "left",
      },
    ],
  })

  for (const disk of disks) {
    table.addRow(diskToRow(disk), { color: rowColor(disk) })
  }

  table.printTable()
}

function prettifyKB(kb: number): string {
  const gb = (kb / 1024 / 1024).toFixed(1)
  return `${gb} G`
}

function parseDisk(line: string): Disk | undefined {
  if (!line.startsWith("/")) return

  const [device, size, used, available, usedPercentage, mount] = line.trim()
    .split(SPACES_REGEX)
  return {
    device,
    used: Number(used),
    usedPercentage: Number(usedPercentage.slice(0, -1)),
    size: Number(size),
    availableKB: Number(available),
    mount,
  }
}

function loadDisks(): Disk[] {
  const p = new Deno.Command("df")
  const { code, stdout, stderr } = p.outputSync()

  if (code !== 0) {
    console.error("Error occured while running 'df'")
    console.error(stderr)
    Deno.exit(code)
  }

  const lines = new TextDecoder().decode(stdout).split("\n")
  return lines
    .map(parseDisk)
    .filter(isPresent)
    .filter((disk) => !EXCLUDE_MOUNT_POINTS.test(disk.mount))
}

function isPresent<T>(value: T | null | undefined): value is T {
  return value != null
}

function warn(disks: Disk[]): void {
  for (const disk of disks) {
    const availableGB = disk.availableKB / 1024 / 1024
    if (availableGB < MIN_FREE_GB) {
      console.warn(
        `Disk "${disk.mount}" has low free space (only ${
          prettifyKB(disk.availableKB)
        })`,
      )
    }
  }
}

main()
